### PR TITLE
Issue #306 : Adding the ability to use overrides

### PR DESCRIPTION
**Pull request by cwilson21**
After working to get this implemented and tested. It appears that the changes I have added are working. 

This creates the `/opt/stackstorm/overrides/` mountpoint in the `st2client` pod so that if reloads happen in the pod it will keep the settings provided by users in the overrides. This also registers overrides when the st2 cluster is being built via the `register-content` job.

THis is just extending the now existing override functionality that exists in Stackstorm 3.7.0 into the HA version of Stackstorm that is now operating on version 3.7.0

```curtis-wilson:test-st2 curtis.wilson$ k -n stackstorm-local logs stackstorm-local-job-st2-register-content--1-xzwx5 -f
2022-05-13 22:08:52,283 INFO [-] Connecting to database "st2" @ "mongodb-local-headless:27017" as user "st2-admin".
2022-05-13 22:08:52,294 INFO [-] Successfully connected to database "st2" @ "mongodb-local-headless:27017" as user "st2-admin".
2022-05-13 22:08:52,734 INFO [-] =========================================================
2022-05-13 22:08:52,734 INFO [-] ############## Registering triggers #####################
2022-05-13 22:08:52,734 INFO [-] =========================================================
2022-05-13 22:08:53,152 INFO [-] Registered 1 triggers.
2022-05-13 22:08:53,152 INFO [-] =========================================================
2022-05-13 22:08:53,152 INFO [-] ############## Registering sensors ######################
2022-05-13 22:08:53,153 INFO [-] =========================================================
2022-05-13 22:08:54,528 INFO [-] Registered 40 sensors.
2022-05-13 22:08:54,529 INFO [-] 0 sensors had their metadata overridden.
2022-05-13 22:08:54,529 INFO [-] =========================================================
2022-05-13 22:08:54,529 INFO [-] ############## Registering runners ######################
2022-05-13 22:08:54,529 INFO [-] =========================================================
2022-05-13 22:08:55,912 INFO [-] Registered 14 runners.
2022-05-13 22:08:55,912 INFO [-] =========================================================
2022-05-13 22:08:55,912 INFO [-] ############## Registering actions ######################
2022-05-13 22:08:55,913 INFO [-] =========================================================
2022-05-13 22:10:24,631 INFO [-] Registered 4473 actions.
2022-05-13 22:10:24,631 INFO [-] 0 actions had their metadata overridden.
2022-05-13 22:10:24,631 INFO [-] =========================================================
2022-05-13 22:10:24,631 INFO [-] ############## Registering rules ########################
2022-05-13 22:10:24,631 INFO [-] =========================================================
2022-05-13 22:10:24,805 INFO [-] Registered 5 rules.
2022-05-13 22:10:24,805 INFO [-] 5 rules had their metadata overridden.
2022-05-13 22:10:24,805 INFO [-] =========================================================
2022-05-13 22:10:24,805 INFO [-] ############## Registering aliases ######################
2022-05-13 22:10:24,805 INFO [-] =========================================================
2022-05-13 22:10:25,081 INFO [-] Registered 15 aliases.
2022-05-13 22:10:25,081 INFO [-] 0 aliases had their metadata overridden.
2022-05-13 22:10:25,082 INFO [-] =========================================================
2022-05-13 22:10:25,082 INFO [-] ############## Registering policy types #################
2022-05-13 22:10:25,082 INFO [-] =========================================================
2022-05-13 22:10:25,112 INFO [-] Registered 3 policy types.
2022-05-13 22:10:25,112 INFO [-] =========================================================
2022-05-13 22:10:25,112 INFO [-] ############## Registering policies #####################
2022-05-13 22:10:25,112 INFO [-] =========================================================
2022-05-13 22:10:25,117 INFO [-] Registered 0 policies.
2022-05-13 22:10:25,117 INFO [-] =========================================================
2022-05-13 22:10:25,117 INFO [-] ############## Registering configs ######################
2022-05-13 22:10:25,118 INFO [-] =========================================================
2022-05-13 22:10:25,158 INFO [-] Registered 4 configs.```

### CODE REVIEW

File: values.yaml
Comment (armab) [inline values.yaml:73]: Would be better to retain the original naming, so it's more obvious for the users:
```suggestion
  overrides:
```

File: values.yaml
Comment (armab) [inline values.yaml:1080]: please keep the newline at the end of file

File: templates/configmaps_overrides.yaml
Comment (armab) [inline templates/configmaps_overrides.yaml:18]: please keep the newline at the end of file

File: values.yaml
Comment (armab) [inline values.yaml:74]: I'm thinking that enabled/disabled logic might be an extra option for overrides.
Unlike RBAC that's enabled/disabled in `st2.conf`, overrides are always on in st2.

Just letting the user specifying files directly might be an easier and sufficient UX for this:
```
st2:
  overrides:
    _global.yaml: |
      ---
      actions:
        defaults:
          enabled: false
    pack1.yaml: |
      # overrides for pack1
    pack2.yaml: |
      # overrides for pack2
```
https://docs.stackstorm.com/latest/packs.html#overriding-pack-defaults

@cwilson21 @cognifloyd WDYT?

File: values.yaml
Comment (armab) [inline values.yaml:72]: Similar to `st2.packs`, it would be also nice to provide a commented example in the Helm values for config overrides so users could configure it more easily.

**Review by armab — CHANGES_REQUESTED** [review]
Thanks for the contribution! A really clean implementation :+1: 

Left several comments to address and discussion around the expected interface.

File: values.yaml
Comment (cwilson21) [reply values.yaml:72]: I can get this added! Not a problem at all (:

File: values.yaml
Comment (cwilson21) [reply values.yaml:74]: So I did the enabled/disabled option as it is a resource that is mounted that may or may not be an additional monthly/hourly charge. I am okay with either way.

File: values.yaml
Comment (cwilson21) [reply values.yaml:73]: I agree, I will get this pushed up!

File: values.yaml
Comment (cognifloyd) [reply values.yaml:74]: If someone wants to disable it, they should just not include any overrides files.
````
st2:
  overrides: {}
```

Comment (cwilson21) [reply values.yaml:74]: Sounds good to me I will make the last changes to it when I get home today and get those into this request !

File: values.yaml
Comment (cwilson21) [reply values.yaml]: This has been updated now!

File: values.yaml
Comment (cwilson21) [reply values.yaml]: New line added!

File: templates/configmaps_overrides.yaml
Comment (cwilson21) [reply templates/configmaps_overrides.yaml]: New line added!

File: values.yaml
Comment (cwilson21) [reply values.yaml:87]: Removed the checked for enabled, and just for if `.Values.st2.overrides.configs` exists

File: values.yaml
Comment (cwilson21) [reply values.yaml]: Added the commented example.

File: values.yaml
Comment (armab) [reply values.yaml:87]: That's close.
 
But instead of the following interface:
```yaml
st2:
  overrides:
    enabled: false
    config:
      _global.yaml: |
      pack1.yaml: |
```

we want to provide a simplified one, with just the override files:
```yaml
st2:
  overrides:
    _global.yaml: |
    pack1.yaml: |
```

and as @cognifloyd said, overrides could be disabled by providing an empty dict instead of `enabled` flag:
```yaml
st2:
  overrides: {}
```

Comment (cwilson21) [reply values.yaml:87]: Ah okay. I misread it no problem, can get this updated should be long

File: templates/deployments.yaml
Comment (cognifloyd) [inline templates/deployments.yaml:1703]: Wrap in an if overrides block.

File: templates/jobs.yaml
Comment (cognifloyd) [inline templates/jobs.yaml:419]: Wrap in an if overrides block

File: values.yaml
Comment (cognifloyd) [inline values.yaml:82]:
```suggestion
  #  packA.yaml: |
  #    ---
  #    rules:
  #      rule.name:
  #        enabled: false
```

File: templates/_helpers.tpl
Comment (cognifloyd) [inline templates/_helpers.tpl:216]:
```suggestion
  {{- if .Values.st2.overrides }}
```

File: templates/_helpers.tpl
Comment (cognifloyd) [inline templates/_helpers.tpl:223]:
```suggestion
  {{- if .Values.st2.overrides }}
```

File: templates/deployments.yaml
Comment (cognifloyd) [inline templates/deployments.yaml:1780]: Can you use your helper template instead here?

File: templates/deployments.yaml
Comment (cognifloyd) [inline templates/deployments.yaml:1836]: Can you use your helper template instead here?

**Review by cognifloyd — COMMENTED** [review]
You've only mounted the overrides under the st2client deployment. So we need this inner any other deployments like st2api? Or st2actionrunner?

And I agree that the register-content job is the only default job that needs the overrides. But, could we also add the overrides mount to the extra_hooks jobs?

Plus a few other comments.
Thanks again for working on this

**Comment by cwilson21** [comment]
I don't think that this needs to be in any other deployment likes `st2api` or `st2actionrunner`. I say that based on I am only adding this to client so that if a `st2ctl reload` is ran it does not nuke the overrides that were put in when register-content happened. As far as I see from my testing being in just `st2-client` is fine, I have a job in my local I was working on previously for our implementation for backing up mongo to `gcs` and that job has not ran since I enacted the global override on rules to set them to `enabled: false` . I appreciate being able to work on this, I am fairly new to working with k8s so this was a fun adventure and I will read through your suggestions and correct what was missed and see about implementing the changes to use my helper template.

**Comment by cognifloyd** [comment]
Also could you add a unit test (something that defines the overrides in values and asserts that the appropriate deployments and jobs Jane the mounts)? Please and thanks!

File: templates/deployments.yaml
Comment (cwilson21) [reply templates/deployments.yaml]: Used the template here.

File: templates/deployments.yaml
Comment (cwilson21) [reply templates/deployments.yaml]: Used the template here

File: templates/deployments.yaml
Comment (cwilson21) [reply templates/deployments.yaml:1704]: wrapped in overrides block

**Comment by cwilson21** [comment]
Added a unit test that checks for the.  mountpoints to exist when overrides are enabled. Sorry for all the commits, I was trying to understand the unit tests and how they work.

File: templates/_helpers.tpl
Comment (cognifloyd) [inline templates/_helpers.tpl:215]: Hmm. Could we rename both of these helpers to use "overrides" instead of "override" (to consistently use the feature's name)?
I know it's nitpicky. Sorry.

**Review by cognifloyd — COMMENTED** [review]
It took me awhile to wrap my head around the unittests as well. Thanks for doing that!

I'm sorry about how nitpicky this is: would you please replace "override" with "overrides" to consistently use the feature's name. I left a few comments where this needs to be fixed, and the new unit test file has this in a bunch of places.

Beyond that, I'm asking @amanda11 about whether we should include the overrides files for any of the services beyond `st2client` and the register-content job. We're almost there. Thank you for all your work on this!

File: templates/configmaps_overrides.yaml
Comment (cwilson21) [reply templates/configmaps_overrides.yaml:6]: Okay so just to condense this into one response. I have made all the changes requested from `override` to `overrides` just waiting for tests to complete in my repo before I merge and it shows up here. No problem working on this for me, again this is something we are actively trying to use in our instances and adding it benefits us and the community!

**Comment by cwilson21** [comment]
Ah I think I missed a few places I will address them shortly

File: tests/unit/overrides_test.yaml
Comment (cognifloyd) [inline tests/unit/overrides_test.yaml:61]:
```suggestion
      - contains: &overrides_volume
          path: spec.template.spec.volumes
          content:
            name: st2-overrides-vol
            configMap:
              name: st2ha-st2-overrides-configs
        documentIndex: 3 # register_content

   
      - contains: &overrides_mnt
          path: spec.template.spec.containers[0].volumeMounts
          content:
            name: st2-overrides-vol
            mountPath: /opt/stackstorm/overrides
        documentIndex: 3 # register_content
```

Comment (cognifloyd) [reply tests/unit/overrides_test.yaml:61]: And add
```
      - notContains: *overrides_volume
        documentIndex: 4 # extra_hooks job
      - notContains: *overrides_mnt
        documentIndex: 4 # extra_hooks job
```

Comment (cwilson21) [reply tests/unit/overrides_test.yaml]: Added `documentIndex: 4` however set it as a contains as one of the requests further up was to add the mount to extra jobs :)

Comment (cognifloyd) [reply tests/unit/overrides_test.yaml]: oops. copy pasta - I meant to change that before submitting :facepalm: Thanks

File: tests/unit/overrides_test.yaml
Comment (cognifloyd) [inline tests/unit/overrides_test.yaml:2]:
```suggestion
suite: Overrides check
```

File: tests/unit/overrides_test.yaml
Comment (cognifloyd) [inline tests/unit/overrides_test.yaml:120]:
```suggestion
      - notContains: *overrides_volume
        documentIndex: 10 # st2actionrunner
      - notContains: *overrides_mnt
        documentIndex: 10 # st2actionrunner
```

File: tests/unit/overrides_test.yaml
Comment (cognifloyd) [inline tests/unit/overrides_test.yaml:27]:
```suggestion
        overrides: #Enabling the overrides mounts in register-content job.
```

**Review by cognifloyd — APPROVED** [review]
Excellent. Thanks for dealing with all my nitpicks.

**Comment by cognifloyd** [comment]
Oh! We also need a changelog entry for this. Please and thank you!

**Comment by cwilson21** [comment]
@cognifloyd what version would we call this? 0.100.0 or is this going to be apart of 0.90.0?

**Comment by cwilson21** [comment]
nvm. Just throwing it in the `In Development` section

**Comment by cognifloyd** [comment]
@armab wdyt? I think this is good to merge.

**Comment by cwilson21** [comment]
Missed one change request for the trailing slash on the mountpoint and got that updated :)

File: templates/_helpers.tpl
Comment (cwilson21) [reply templates/_helpers.tpl]: Adjusted this

File: values.yaml
Comment (armab) [inline values.yaml:78]: Looks like a wrong indentation here for `packA.yaml`, compared to `_global.yaml`.

File: values.yaml
Comment (armab) [inline values.yaml:82]: I'd also prefer if we'll place this commented example after the `overrides: {}` line.
That would be consistent with the other examples in Helm values.

**Review by armab — APPROVED** [review]
Left a couple small comments,

besides that LGTM :+1:

File: values.yaml
Comment (cognifloyd) [reply values.yaml:82]:
```suggestion
  #  overrides:
  #    _global.yaml: |
  #      ---
  #      rules:
  #        defaults:
  #          enabled: true
  #    packA.yaml: |
  #      ---
  #      rules:
  #        rule.name:
  #          enabled: false
```

Comment (cwilson21) [reply values.yaml]: Made the changes here to place the example below `overrides`

File: values.yaml
Comment (cwilson21) [reply values.yaml]: The indentation should be fixed now.

**Comment by armab** [comment]
Nice!
That's a high-quality contribution Thanks, everyone :+1:

**Pull request merged**
